### PR TITLE
App Configuration Property cache

### PR DESCRIPTION
**Pull request by mrm9084**
Adds a Cache to App Configuration that enables etag checking on an individual level along with the Config Store as a whole. Allows for individual Settings to update without also pulling in all Settings again.

### CODE REVIEW

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java [hunk 1]
```diff
@@ -98,16 +112,44 @@ private void refresh(ConfigStore store, String storeSuffix, String watchedKeyNam
...
+                    storeEtagMap.put(storeName, etag);
+                    options = new QueryOptions().withKeyNames(refreshKey)
+                            .withLabels(store.getLabels()).withFields(QueryField.ETAG).withRange(0, 0);
+
+                    keyValueItems = configOperations.getRevisions(store.getName(), options);
+
+                    if (keyValueItems.isEmpty() || keyValueItems.get(0).getEtag()
+                            .equals(propertyCache.getCachedEtag(refreshKey))) {
+                        refreshKeys = propertyCache.updateRefreshCacheTimeForKey(store.getName(), refreshKey, date);
+                        i--;
```
Comment (sophiaso) [inline spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java]: The `i--` here sounds special, can we add some comments? Will it make `i` to `-1`?

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java [hunk 2]
```diff
@@ -98,16 +112,44 @@ private void refresh(ConfigStore store, String storeSuffix, String watchedKeyNam
...
         }
+        
+        if (!etag.equals(storeEtagMap.get(storeName))) {
+            Date date = new Date();
+
+            // Checks all cached items to see if they have been updated
+            List<String> refreshKeys = propertyCache.getRefreshKeys(store.getName());
+            for (int i = 0; i < refreshKeys.size(); i++) {
+                String refreshKey = refreshKeys.get(i);
+                if (refreshKey.contains(watchedKeyNames.replace("*", ""))) {
```
Comment (sophiaso) [inline spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java]: Just question, the contains compare is different with regexp match, though regexp much requires some performance cost, will this contain match cases not expected? If yes, or we tolerate such match?

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java [hunk 3]
```diff
@@ -109,8 +110,14 @@ public ConfigServiceOperations azureConfigOperations(ConfigHttpClient client, Co
...
     public AzureConfigPropertySourceLocator sourceLocator(ConfigServiceOperations operations,
-                                                          AzureCloudConfigProperties properties) {
-        return new AzureConfigPropertySourceLocator(operations, properties);
+            AzureCloudConfigProperties properties, PropertyCache propertyCache) {
+        return new AzureConfigPropertySourceLocator(operations, properties, propertyCache);
+    }
+    
+    @Bean
+    @ConditionalOnMissingBean
+    public PropertyCache getPropertyCache() {
```
Comment (sophiaso) [inline spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java]: nit: why we need to define the `PropertyCache` bean twice? Another one is in `AzureCloudConfigAutoConfiguration.java `.

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java [hunk 4]
```diff
@@ -65,22 +69,90 @@ public Object getProperty(String name) {
...
+            QueryOptions queryOptions = new QueryOptions().withKeyNames(context + "*").withLabels(label);
+            List<KeyValueItem> items = source.getKeys(storeName, queryOptions);
+            for (KeyValueItem item : items) {
+                String key = item.getKey().trim().substring(context.length()).replace('/', '.');
+                properties.put(key, item.getValue());
+            }
+            propertyCache.addKeyValuesToCache(items, storeName, date);
+
+            // Reading In Features
+            queryOptions = new QueryOptions().withKeyNames("*appconfig*").withLabels(label);
```
Comment (sophiaso) [inline spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java]: nit: I think we already have this constant defined somewhere?

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java (hunk 3 above)
Comment (mrm9084) [reply spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java]: Removed duplicate in `AzureCloudConfigAutoConfiguration.java`

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java [hunk 5]
```diff
@@ -65,22 +69,90 @@ public Object getProperty(String name) {
...
+                    properties.put(cachedKey, propertyCache.getCachedValue(key));
+                } else {
+                    List<KeyValueItem> items = new ArrayList<KeyValueItem>();
+                    KeyValueItem item = new KeyValueItem();
+                    item.setKey(key);
+                    item.setValue(propertyCache.getCachedValue(key));
+                    item.setContentType(FEATURE_FLAG_CONTENT_TYPE);
+                    items.add(item);
+
+                    createFeatureSet(items, propertyCache, date);
```
Comment (saragluna) [inline spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java]: The updated features will not go to properties?

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java [hunk 6]
```diff
@@ -98,16 +112,47 @@ private void refresh(ConfigStore store, String storeSuffix, String watchedKeyNam
...
-            return;
+            propertyCache.updateRefreshCacheTime(store.getName(), watchedKeyNames, delay);
         }
+        
+        if (!etag.equals(storeEtagMap.get(storeName))) {
+            Date date = new Date();
+            String watchedKeyNamesPrefix = watchedKeyNames.replace("*", "");
+
+            // Checks all cached items to see if they have been updated
+            List<String> refreshKeys = propertyCache.getRefreshKeys(store.getName());
```
[the remark this answers is not in the document]
Comment (mrm9084) [reply spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java]: Renaming storeName to storeNameWithPrefix for clarification.

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceLocator.java [hunk 7]
```diff
@@ -126,15 +148,29 @@ private String propWithProfile(String context, String profile) {
...
+            Map<String, List<String>> storeContextsMap, boolean initFeatures) throws IOException {
         List<AzureConfigPropertySource> sourceList = new ArrayList<>();
 
         for (String label : store.getLabels()) {
             AzureConfigPropertySource propertySource = new AzureConfigPropertySource(context, operations,
                     store.getName(), label, properties);
 
-            propertySource.initProperties();
+            propertySource.initProperties(propertyCache);
+            if (initFeatures) {
```
[the remark this answers is not in the document]
Comment (mrm9084) [reply spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceLocator.java]: We need all of the feature flags from the various confg stores to be in the same Property Source so they don't overwrite each other as they are set to a single value. This way the last config store will set all feature flags found.

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java (hunk 5 above)
Comment (mrm9084) [reply spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java]: That is the intent see comment about "distinguish features".

**Pull request merged**